### PR TITLE
Add index_reduce decomposition

### PR DESCRIPTION
**Pull request by pearu**
As in the title.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122579
* #123375

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/decomposition.py [hunk 1]
```diff
@@ -680,3 +680,25 @@ def put(self, index, source, accumulate=False):
 def put_(self, index, source, accumulate=False):
     out = aten.put(self, index, source, accumulate=accumulate)
     return self.copy_(out)
+
+
+@register_decomposition(aten.index_reduce)
+def index_reduce(self, dim: int, index, src, reduction_type: str, **kwargs):
+    return index_reduce_(self.clone(), dim, index, src, reduction_type, **kwargs)
```
Comment (justinchuby) [inline torch/_inductor/decomposition.py:687]: Is this decomposition decomposing a functional op with a inplace op? Would it be desirable to decompose it with a functional op as well?

File: torch/_inductor/decomposition.py [hunk 2]
```diff
@@ -680,3 +680,34 @@ def put(self, index, source, accumulate=False):
...
+
+@register_decomposition(aten.index_reduce)
+def index_reduce(
+    self, dim: int, index, src, reduction_type: str, *, include_self: bool = True
+):
+    repeats = self.shape[dim + 1 :].numel() * self.shape[:dim].numel()
+    index_shape = (index.numel(), *self.shape[dim + 1 :], *self.shape[:dim])
+    perm = (*range(self.ndim - dim, self.ndim), 0, *range(1, self.ndim - dim))
+    scatter_index = index.repeat_interleave(repeats).reshape(index_shape).permute(perm)
+    return self.scatter_reduce(
```
Comment (peterbell10) [inline torch/_inductor/decomposition.py:693]: For certain inputs `scatter_reduce` falls back to eager, in which case this ends up creating a massive intermediate index tensor which isn't ideal. 
https://github.com/pytorch/pytorch/blob/c7bf5871ceb02321ec8d8fd5fd305ad9c2fd8132/torch/_inductor/lowering.py#L3420-L3431
Thankfully for inductor-specific decompositions you can conditionally not decompose by returning `NotImplemented`. So I would refactor `scatter_fallback` to extract the conditions under which it falls back and not decompose in those cases.

File: torch/_inductor/decomposition.py [hunk 3]
```diff
@@ -680,3 +680,34 @@ def put(self, index, source, accumulate=False):
...
+    self, dim: int, index, src, reduction_type: str, *, include_self: bool = True
+):
+    out = self.index_reduce(
+        dim,
+        index,
+        src,
+        reduction_type,
+        include_self=include_self,
+    )
+    return self.copy_(out)
```
Comment (peterbell10) [inline torch/_inductor/decomposition.py:713]: I think this should already be covered by this existing decomp:
https://github.com/pytorch/pytorch/blob/c7bf5871ceb02321ec8d8fd5fd305ad9c2fd8132/torch/_decomp/decompositions.py#L4783

**Review by peterbell10 — CHANGES_REQUESTED** [review]

File: torch/_inductor/decomposition.py (hunk 3 above)
Comment (pearu) [reply torch/_inductor/decomposition.py]: Right, thanks for pointing it out!

File: torch/_inductor/decomposition.py (hunk 2 above)
Comment (pearu) [reply torch/_inductor/decomposition.py:704]: Thanks for the hint! I have refactored parts of `scatter_fallback` into `use_scatter_fallback` that is used both in lowering and decomposition of relevant ops.

**Comment by pearu** [comment]
> Test failure is real, and it would also be nice to see the generated code and a benchmark to confirm it's not regressing performance.

The test failure is now fixed.

Considering the fact that `index_reduce` decomposition relies on the decomposition of `scatter_reduce`, and how restrictive is `use_scatter_fallback(...)` to disable `scatter_reduce` fallback, `index_reduce` will always fallback to its aten implementation (notice that the expression `reduction_type not in {None, reduce_ty}` in `use_scatter_fallback` is always True when called from `index_reduce` because `reduction_type` is in `{"prod", "mean", "amax", "amin"}`). So, I think enabling `index_reduce` decomposition will be blocked by `scatter_reduce` that in most cases falls back to its aten implementation.

**Comment by peterbell10** [comment]
Hrm that's a bit disappointing. triton does have `atomic_min` and `atomic_max` but now you mention it these aren't exposed in the inductor ir. There is `ops.store(x, mode="atomic_add")` which could be extended to support min and max though.

Also, looking at how `reduce="mean"` is implemented, I think this could easily be added to your decomposition:
https://github.com/pytorch/pytorch/blob/9b90c5e2a12afb58f19c55a74af60150b4cc03e8/aten/src/ATen/native/cuda/Indexing.cu#L1195-L1203
`index_reduce_func_cuda_impl` is actually just doing a sum here and we divide by the count in separate kernels.

**Comment by pearu** [comment]
@peterbell10 so, IIUC, you are suggest not relying on `scatter_reduce` but implementing `index_reduce` decomposition from scratch. Is this correct?

**Comment by peterbell10** [comment]
I'm suggesting that:
1. `index_reduce` with `reduce="mean"` could be decomposed into `index_add` in this PR, similar to the CUDA implementation above
2. In a future PR, supporting `scatter_reduce_` with "amax" and "amin" reductions should be possible by extending `ops.store` to support `atomic_max` and `atomic_min` as mode arguments. This would automatically make this `index_reduce` decomposition more useful.

**Comment by pearu** [comment]
> I'm suggesting that:
> 
> 1. `index_reduce` with `reduce="mean"` could be decomposed into `index_add` in this PR, similar to the CUDA implementation above

Done.

File: torch/_inductor/decomposition.py [hunk 4]
```diff
@@ -705,3 +706,50 @@ def _softmax_backward_data(grad_output, output, dim, input_dtype):
...
+    ):
+        tmp = (torch.ones_like if include_self else torch.zeros_like)(self)
+        counts = tmp.index_add(dim, index, torch.ones_like(src))
+        out = self if include_self else tmp  # tmp is torch.zeros_like(self)
+        out = out.index_add(dim, index, src)
+        mask = counts == 0
+        if self.dtype.is_floating_point:
+            out = out / counts
+        else:
+            out = out // counts.masked_fill(mask, 1)
```
Comment (peterbell10) [inline torch/_inductor/decomposition.py:726]: Why do you only include the masked fill in the non-fp case? The C++ version has it outside of the if, so it applies to both cases.

Comment (pearu) [reply torch/_inductor/decomposition.py:726]: First, `mask = counts == 0` is needed only when `include_self==False` (in the `include_self==True` case, minimal count is always 1 or greater).

In the fp-case, there is no need for the `masked_fill` op because true-division (`/`) always succeeds. The `out` entries that have zero counts will be replaced by the corresponding `self` entries (see `where` expression below) and the results from true-zero-division events will be discarded anyway. In this regard, the C++ version is sub-optimal.

In the non-fp-case, floor division (`//`) would throw when dividing by zero, hence the need for the `masked_fill` to protect from floor-zero-division events.

Comment (peterbell10) [reply torch/_inductor/decomposition.py:726]: > First, mask = counts == 0 is needed only when include_self==False (in the include_self==True case, minimal count is always 1 or greater).

That makes sense, but in that case why mask for `include_self=True` on non-fp? Instead you could do:
```python
if not include_self:
    counts = counts.masked_fill(count == 0, 1)

return out / counts if self.dtype.is_floating_point else out // counts
```

File: torch/_inductor/decomposition.py [hunk 5]
```diff
@@ -705,3 +706,50 @@ def _softmax_backward_data(grad_output, output, dim, input_dtype):
...
+def index_reduce(
+    self, dim: int, index, src, reduction_type: str, *, include_self: bool = True
+):
+    if reduction_type == "mean" and not needs_fallback_due_to_atomic_add_limitations(
+        self.dtype
+    ):
+        tmp = (torch.ones_like if include_self else torch.zeros_like)(self)
+        counts = tmp.index_add(dim, index, torch.ones_like(src))
+        out = self if include_self else tmp  # tmp is torch.zeros_like(self)
+        out = out.index_add(dim, index, src)
```
Comment (peterbell10) [inline torch/_inductor/decomposition.py:721]: I think the accuracy failures are because we need to zero out the indexed locations before calling `index_add` in the the `include_self=False` case:
https://github.com/pytorch/pytorch/blob/5a15cbfa440e87f8439d1f28b3d8df0c51daf088/aten/src/ATen/native/cuda/Indexing.cu#L1026-L1050

File: torch/_inductor/decomposition.py (hunk 4 above)
Comment (pearu) [reply torch/_inductor/decomposition.py:726]: The suggested code has two problems:
- it is incomplete as it lacks initializing `out` with `self` entries when `counts==0` is True and `include_self=False`
- it is less optimal as calling `masked_fill` is superfluous when `self.dtype.is_floating_point` is True

File: torch/_inductor/decomposition.py (hunk 5 above)
Comment (pearu) [reply torch/_inductor/decomposition.py:721]: Hmm, in the `include_self=False` case, `out` is defined as `torch.zeros_like(self)` (see the definition of `tmp`). IIUC, `out` is already initialized with zeros. So, I don't see a problem here unless calling `tmp.index_add` has side-effects to `tmp` data.

Comment (pearu) [reply torch/_inductor/decomposition.py:721]: There appears to be a bug that leads to these failures. For instance, for inputs with `shape=(1,)` and having `counts=[1]`, both `mask = counts == 0` and `mask = counts == 1` lead to the same result: `mask == [True]`.

Comment (peterbell10) [reply torch/_inductor/decomposition.py:721]: I don't think out should be initialized with zeros when `include_self = False`, I can't see that in the eager code.

Comment (peterbell10) [reply torch/_inductor/decomposition.py:721]: Also how did you inspect the mask? I wouldn't expect that to be a real buffer in the generated code.

Comment (pearu) [reply torch/_inductor/decomposition.py:721]: > I don't think out should be initialized with zeros when include_self = False, I can't see that in the eager code.

`out` is initialized with zeros in
https://github.com/pytorch/pytorch/blob/5f378e1853558d603028c2062373869c6f192885/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L1016

Comment (pearu) [reply torch/_inductor/decomposition.py:721]: Anyway, using
```
mask = counts < 1
```
fixed the issue (tested locally).

Comment (peterbell10) [reply torch/_inductor/decomposition.py:721]: Yes precisely, `index_fill` not `fill_` so it's not the whole tensor that's zeroed, only the indexes specified.

File: torch/_inductor/decomposition.py (hunk 4 above)
Comment (peterbell10) [reply torch/_inductor/decomposition.py:726]: > it is less optimal as calling masked_fill is superfluous when self.dtype.is_floating_point is True

The `where` in your code is only needed because you don't `masked_fill` for float.

File: torch/_inductor/decomposition.py (hunk 5 above)
Comment (pearu) [reply torch/_inductor/decomposition.py:721]: That is a very good point, thanks! When using `index_fill`, the `index_reduce` implementation will simplify.

**Review by peterbell10 — COMMENTED** [review]
Looks good, just needs a benchmark to confirm

**Comment by pearu** [comment]
> Looks good, just needs a benchmark to confirm

```
[ index_select((30, 40, 50)-tensor, 0, (1000,)-tensor, (1000, 40, 50)-tensor, mean) ]
                                               |  Eager  |  Decomposed
18 threads: ----------------------------------------------------------
      torch.int32 cuda:0 include_self=False    |   150   |       90   
      torch.int32 cuda:0 include_self=True     |   124   |       80   
      torch.int64 cuda:0 include_self=False    |   193   |      193   
      torch.int64 cuda:0 include_self=True     |   167   |      170   
      torch.float32 cuda:0 include_self=False  |   149   |       90   
      torch.float32 cuda:0 include_self=True   |   123   |       80   
      torch.float64 cuda:0 include_self=False  |   193   |      137   
      torch.float64 cuda:0 include_self=True   |   167   |      119   

Times are in microseconds (us).
```

**Review by peterbell10 — APPROVED** [review]
